llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2449,8 +2489,39 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpStore>(SPIRVValue *const
 
   SPIRVType *const spvStoreType = spvStore->getDst()->getType();
 
-  addStoreInstRecursively(spvStoreType->getPointerElementType(), storePointer, storeValue, isVolatile, isCoherent,
-                          isNonTemporal);
+  SPIRVToLLVM::ScratchBoundsCheckInfo boundsCheckInfo = prepareScratchBoundsCheck(spvStore->getDst());
+
+  if (boundsCheckInfo.shouldInsertBoundsCheck) {
+    // An OpStore will be converted to a chain of comparison results against all bounds in the OpStore.
+    // The actual work (adding the store to the store block) and just omitting the store if the OOB check failed is done
+    // here:
+
+    // .parentBlock:
+    // foreach (indexToAccess in accessChain)
+    // addUlt(indexToAccess, accessChain.relevantArray);
+    // jmpeq .storeBlock, !oob ; else branch to final block
+
+    // .storeBlock:
+    // store...
+    // jmp .finalBlock
+
+    // .finalBlock:
+    // ; ... all subsequent instructions
+    getBuilder()->CreateCondBr(boundsCheckInfo.finalCmpResult, boundsCheckInfo.intermediateBlock,
+                               boundsCheckInfo.finalBlock);
+
+    // Add the store instruction to the load block, followed by an unconditional jump.
+    getBuilder()->SetInsertPoint(boundsCheckInfo.intermediateBlock);
+    addStoreInstRecursively(spvStoreType->getPointerElementType(), storePointer, storeValue, isVolatile, isCoherent,
```

Comment:
Can you hoist `spvStoreType->getPointerElementType()` into a variable? It's used in both then and else, and it would make it easier to print it in the debugger.